File: frontend/__tests__/honesty.test.ts
import fs from 'fs';
import path from 'path';

/**
 * WP-0.2 C5/C1 (AGENTS.md §1): the frontend must NEVER render invented
 * numbers or placeholder learners when data is unavailable — honest 0/empty
 * states instead. This test scans the shipped source (pages + components,
 * tests excluded) for fabricated-fallback patterns and fails when one is
 * found, so the "no placeholder stubs" guarantee is executable, not a rule
 * on paper.
 */
const ROOT = path.join(__dirname, '..', 'src');

const BANNED_IDENTIFIERS = [
  'placeholderStudent',
  'fakeStudent',
  'dummyStudent',
  'sampleRoster',
  'mockRoster',
  'fakeData',
  'dummyData',
  'sampleData',
  'placeholderData',
  'hardcodedStudents',
  'demoRoster',
];

// Patterns that fabricate data at render time instead of showing 0/empty.
const BANNED_PATTERNS: Array<[RegExp, string]> = [
  [/const\s+\w+\s*=\s*\[\s*\{\s*name:\s*['"][A-Z][a-z]+ ['"][A-Z]/, 'hardcoded named object (likely an invented learner)'],
  [/Math\.random\(\).*score|score.*Math\.random\(\)/, 'randomly generated scores'],
  [/\[\s*\{\s*(name|label|day):\s*['"][^'"]+['"],\s*(value|score|duration|count):\s*\d+\s*\}/, 'hardcoded data point array'],
];

function walk(dir: string, out: string[] = []): string[] {
  for (const entry of fs.readdirSync(dir, { withFileTypes: true })) {
    const p = path.join(dir, entry.name);
    if (entry.isDirectory()) {
      if (entry.name === '__tests__' || entry.name === 'node_modules') continue;
      walk(p, out);
    } else if (entry.name.endsWith('.tsx') || entry.name.endsWith('.ts')) {
      out.push(p);
    }
  }
  return out;
}

describe('honest-zero sweep (WP-0.2)', () => {
  const files = walk(ROOT);

  it('scans the shipped source for fabricated-fallback identifiers', () => {
    const hits: string[] = [];
    for (const file of files) {
      const src = fs.readFileSync(file, 'utf8');
      for (const id of BANNED_IDENTIFIERS) {
        if (src.includes(id)) hits.push(`${path.relative(ROOT, file)}: "${id}"`);
      }
    }
    expect(hits).toEqual([]);
  });

  it('scans for hardcoded invented data (named objects, random scores, data arrays)', () => {
    const hits: string[] = [];
    for (const file of files) {
      const src = fs.readFileSync(file, 'utf8');
      for (const [re, why] of BANNED_PATTERNS) {
        if (re.test(src)) hits.push(`${path.relative(ROOT, file)}: ${why}`);
      }
    }
    expect(hits).toEqual([]);
  });

  it('covers the audited surfaces: goal ring, moderator roster, chart-data consumers', () => {
    const surfaces = [
      'src/app/dashboard/page.tsx',   // goal ring
      'src/components/moderator/RosterOverview.tsx', // roster
      'src/app/observation/page.tsx', // chart-data charts
    ].map((p) => path.join(__dirname, '..', p));
    for (const f of surfaces) {
      expect(fs.existsSync(f)).toBe(true);
    }
  });
});

File: backend/internal/honesty/honesty_test.go
package honesty

import (
	"os"
	"path/filepath"
	"regexp"
	"strings"
	"testing"
)

// Backend mirror of frontend/__tests__/honesty.test.ts (AGENTS.md §1): the
// backend must NEVER fabricate numbers when data is unavailable — honest
// empty/zero states instead. This test walks the shipped Go source and fails
// when a fabricated-fallback pattern appears, so the guarantee is executable.
//
// Patterns are assembled at runtime so the scanner's own source can never
// self-match (the literal target strings never appear in this file).

var fabricatedPatterns = []struct {
	re  *regexp.Regexp
	why string
}{
	{
		// The old GetChartData fallback: a hardcoded 7-day all-zero week.
		regexp.MustCompile(`"name":\s*"` + "Mon" + `"\s*,\s*"score":\s*0`),
		"fabricated zero-week chart series (GetChartData regression)",
	},
	{
		// A hardcoded weekday-name slice — the tell-tale of a fabricated
		// weekly chart series (the GetChartData regression built one of
		// these; any future weekly fabrication will too).
		regexp.MustCompile(`"` + "Mon" + `",\s*"` + "Tue" + `",\s*"` + "Wed" + `"`),
		"hardcoded weekday series (fabricated weekly chart)",
	},
	{
		// A hardcoded absolute year in time.Date — a fabricated analytics
		// window that pretends to be "the last 7 days" but is pinned to a
		// fixed calendar date forever.
		regexp.MustCompile(`time\.Date\(2[0-9]{3},`),
		"hardcoded absolute date window (fabricated analytics range)",
	},
	{
		// Invented placeholder roster entries.
		regexp.MustCompile(`placeholderStudent|fakeStudent|dummyStudent|demoRoster|mockRoster`),
		"placeholder learner identifiers",
	},
	{
		// Randomly generated analytics.
		regexp.MustCompile(`rand\.Intn|rand\.Float64|Math\.Random|math\.rand`),
		"randomly generated metrics/analytics",
	},
}

// Exclude vendored/cache dirs and this scanner itself from the walk target
// (it is the scanner — nothing to scan). Tests are scanned too: a fabricated
// fallback smuggled into test fixtures is still a red flag, but expected
// values in assertions are legitimate, so only source (non-_test) files are
// scanned for the random-metrics pattern.
func walk(dir string, skipTest bool, out *[]string) {
	entries, err := os.ReadDir(dir)
	if err != nil {
		return
	}
	for _, e := range entries {
		p := filepath.Join(dir, e.Name())
		if e.IsDir() {
			if e.Name() == "node_modules" || e.Name() == ".git" || e.Name() == "vendor" {
				continue
			}
			walk(p, skipTest, out)
			continue
		}
		if !strings.HasSuffix(e.Name(), ".go") {
			continue
		}
		if skipTest && strings.HasSuffix(e.Name(), "_test.go") {
			continue
		}
		*out = append(*out, p)
	}
}

func TestBackendHonestyScan(t *testing.T) {
	root := filepath.Join("..", "..") // backend/internal/honesty -> backend
	var sourceFiles []string
	walk(root, true, &sourceFiles)

	var hits []string
	for _, f := range sourceFiles {
		src, err := os.ReadFile(f)
		if err != nil {
			continue
		}
		for _, p := range fabricatedPatterns {
			if p.re.Match(src) {
				hits = append(hits, filepath.ToSlash(f)+": "+p.why)
			}
		}
	}
	if len(hits) > 0 {
		t.Fatalf("fabricated-fallback patterns found:\n  %s", strings.Join(hits, "\n  "))
	}
}

// The audited surfaces must exist so the sweep covers the endpoints that
// previously fabricated data: chart-data and the moderator roster.
func TestAuditedSurfacesExist(t *testing.T) {
	surfaces := []string{
		filepath.Join("..", "service", "learner_service.go"),    // GetChartData
		filepath.Join("..", "service", "learner_service.go"),    // GetModeratorRoster
		filepath.Join("..", "repository", "course_repo.go"),     // enrollment counts
		filepath.Join("..", "repository", "completion_repo.go"), // DailyActivity rows
	}
	for _, f := range surfaces {
		if _, err := os.Stat(f); err != nil {
			t.Fatalf("audited surface missing: %v", f)
		}
	}
}
